Lib/asyncio/tasks.py

```diff
@@ -439,63 +440,23 @@ async def wait_for(fut, timeout):
 
     This function is a coroutine.
     """
-    loop = events.get_running_loop()
-
     if timeout is None:
         return await fut
 
     if timeout <= 0:
-        fut = ensure_future(fut, loop=loop)
+        fut = ensure_future(fut)
 
         if fut.done():
             return fut.result()
```

Comment:
It works even without these two lines. :-)

However, this whole block (L443-L456) is somehow required, otherwise several things go wrong. I wonder if that points to a flaw in `timeout()`?

Reply:
Not a flaw but a different design choice. I had pointed this out in https://github.com/python/cpython/pull/31394#discussion_r817952872 on the PR adding `timeout`.

TL;DR
- `asyncio.wait_for` raises `TimeoutError` if `0` or `None` is passed as deadline.
- `asyncio.timeout` disables deadline if `None` is passed.

Anyway changing this is out of scope for this so marking resolved.

Reply:
Correction, wait_for() also disables deadline if None is passed as timeout. It's the first two lines of code in this function. So as far as these two bullets are concerned they are actually equivalent.

So I think there's something else that this block (L443-L456) ensures that timeout() doesn't.

Reply:
Right, the only special case is `<=0` as deadline in `wait_for` which behaves differently. `test_wait_for_timeout_less_then_0_or_0_coroutine_do_not_started` fails if this special case is removed.

Reply:
Removed `None` special case.

Reply:
> Right, the only special case is `<=0` as deadline in `wait_for` which behaves differently. `test_wait_for_timeout_less_then_0_or_0_coroutine_do_not_started` fails if this special case is removed.

Ah, thanks for explaining. Using `wait_for(t, 0)`, `t` is cancelled before it is started, whereas when using `async with timeout(0)` a task is allowed to run until its first `await` or exit.

Now I'm wondering which design is better, but we should first make `wait_for()` work as it always did.